src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -1882,7 +1904,9 @@ private ReadOnlyCollection<TKey> GetKeys()
             try
             {
                 AcquireAllLocks(ref locksAcquired);
-                List<TKey> keys = new List<TKey>();
+
+                int count = GetCountInternal();
+                List<TKey> keys = new List<TKey>(count);
```

Comment:
Or `ArrayBuilder<TKey>`. That still avoids the allocation, but does not require manual tracking of the array index.